### DISCUSSION TITLE
How to add a custom dictionary to a VSCode Workspace project?

### DISCUSSION THREAD

**Discussion by Yohanna** [streetsidesoftware/vscode-spell-checker#2849] [Q&A] [answered]
Hello, I'd like to add a custom dictionary to a [Multi-root Workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces). My folder structure is like this:
- Main workspace directory located at `/home/your-username/your-project`.
- `your-project` contains two folders, `client` and `server`.

I want to create a custom dictionary at the root Workspace folder, so it'll be `/home/your-username/your-project/custom-dictionary.txt`.

Here's my VSCode Workspace `.code-workspace` file contents:

```json
{
  "folders": [
    {
      "path": "client"
    },
    {
      "path": "server"
    }
  ],
  "settings": {
    "cSpell.language": "en,en-GB,en-US",
    "cSpell.diagnosticLevel": "Warning",
    "cSpell.customDictionaries": {
      "custom-dictionary": {
        "name": "Workspace Dictionary",
        "description": "Custom dictionary",
        "path": "${workspaceFolder}/custom-dictionary.txt",
        "addWords": true,
        "scope": "workspace"
      }
    }
  }
}

```

However, this didn't work. CSpell end up creating a custom dictionary at `/home/your-username/your-project/client`, the *first* folder defined in the Workspace. It also throws this error in VSCode logs:

```
uncaughtException Error: /home/your-username/your-project/client/custom-dictionary.txt does not exist.
    at /Users/<user_name>/.vscode/extensions/streetsidesoftware.code-spell-checker-3.0.1/packages/_server/dist/main.cjs:99:99816
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)

```

The only way I could get this to work is if define a *third* folder in my Workspace confingurations for the root Workspace folder, e.g.:

```json
{
  "folders": [
    {
      "path": "client"
    },
    {
      "path": "server"
    },
    {
      "name": "Root",     // <---------  Extra folder
      "path": "./"
    }
  ],
  "settings": {
    "cSpell.language": "en,en-GB,en-US",
    "cSpell.diagnosticLevel": "Warning",
    "cSpell.customDictionaries": {
      "custom-dictionary": {
        "name": "Workspace Dictionary",
        "description": "Custom dictionary",
        "path": "${workspaceFolder:Root}/custom-dictionary.txt",   // <------- Notice the added "Root"
        "addWords": true,
        "scope": "workspace"
      }
    }
  }
}
```
Is this the proper way to add a custom Workspace dictionary, or is this a bug in the extension? The [docs](https://streetsidesoftware.com/vscode-spell-checker/docs/configuration/#cspellcustomdictionaries) and this [comment](https://github.com/streetsidesoftware/vscode-spell-checker/issues/681#issuecomment-950581677) doesn't mention anything about defining a _root_ folder in your Workspace.

**Comment by Jason3S** [comment, marked as answer]
@Yohanna,

It is common for one of the folders in a multi-root workspace to be the "root" and it is generally the first. But, a multi-root workspace does not have to have a root, so the extension has to guess.

If you leave the name off, it looks at the first workspace.

```jsonc
  "settings": {
    "cSpell.language": "en,en-GB,en-US",
    "cSpell.diagnosticLevel": "Warning",
    "cSpell.customDictionaries": {
      "custom-dictionary": {
        "name": "Workspace Dictionary",
        "description": "Custom dictionary",
        "path": "${workspaceFolder:client}/../custom-dictionary.txt",   // <------- You can make it relative to the client or server.
        "addWords": true,
        "scope": "workspace"
      }
    }
  }
```

**Reply by Yohanna** [reply]
Thanks, Jason!
Does the extension has a way of guessing the top-level Workspace directory on its own? Basically where the `.code-workspace` file is located. I think this would avoid any surprises when creating a Workspace scoped dictionary file.
